packages/react-reconciler/src/ReactFiberCommitWork.old.js

```diff
@@ -286,7 +286,33 @@ function safelyAttachRef(current: Fiber, nearestMountedAncestor: Fiber | null) {
 
 function safelyDetachRef(current: Fiber, nearestMountedAncestor: Fiber | null) {
   const ref = current.ref;
-  if (ref !== null) {
+  const refCleanup = current.refCleanup;
+
+  if (refCleanup !== null) {
+    if (typeof ref === 'function') {
+      try {
+        if (shouldProfile(current)) {
+          try {
+            startLayoutEffectTimer();
+            refCleanup();
+          } finally {
+            recordLayoutEffectDuration(current);
+          }
+        } else {
+          refCleanup();
+        }
+      } catch (error) {
+        captureCommitPhaseError(current, nearestMountedAncestor, error);
+      } finally {
+        // `refCleanup` has been called. Nullify all references to it to prevent double invocation.
+        current.refCleanup = null;
+        const finishedWork = current.alternate;
+        if (finishedWork != null) {
+          finishedWork.refCleanup = null;
+        }
+      }
+    }
+  } else if (ref !== null) {
     if (typeof ref === 'function') {
```

Comment:
You can place the `if (refCleanup !== null) {` inside here since you'll always do that check anyway so there's no extra checks. That way when it's not a function but a `ref.current` you don't do extra checks.